internal/backend/runtime/omni/controllers/reporting/reporting.go

```diff
@@ -0,0 +1,214 @@
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/client/pkg/panichandler"
+)
+
+var (
+	// DebounceDuration defines how long to wait after the last event before processing updates.
+	DebounceDuration = 10 * time.Minute
+)
+
+// StripeMetricsReporterControllerName is the name of the StripeMetricsReporterController.
+const StripeMetricsReporterControllerName = "StripeMetricsReporterController"
+
+// StripeMetricsReporterController ...
+type StripeMetricsReporterController struct {
+	generic.NamedController
+}
+
+// NewStripeMetricsReporterController initializes StripeMetricsReporterController.
+func NewStripeMetricsReporterController() *StripeMetricsReporterController {
+	return &StripeMetricsReporterController{
+		NamedController: generic.NamedController{
+			ControllerName: StripeMetricsReporterControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *StripeMetricsReporterController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: resources.DefaultNamespace,
+			Type:      omni.MachineStatusMetricsType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Set your Stripe API key from the environment variable.
+	stripe.Key = os.Getenv("STRIPE_API_KEY")
+	if stripe.Key == "" {
+		return fmt.Errorf("STRIPE_API_KEY environment variable not set")
+	}
+
+	// Retrieve Subscription Item ID from environment variables or configuration.
+	subscriptionItemID := os.Getenv("STRIPE_SUBSCRIPTION_ITEM_ID")
+	if subscriptionItemID == "" {
+		return fmt.Errorf("STRIPE_SUBSCRIPTION_ITEM_ID environment variable not set")
+	}
+
+	// Channel to send count changes to the debouncer.
+	countChan := make(chan uint32, 100)
+
+	// Start the debouncer goroutine.
+	panichandler.Go(func() {
+		debouncer(ctx, countChan, subscriptionItemID)
+	}, nil)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			metrics, err := safe.ReaderGetByID[*omni.MachineStatusMetrics](ctx, r, omni.MachineStatusMetricsID)
+			if err != nil {
+				if state.IsNotFoundError(err) {
+					continue
+				}
+
+				return err
+			}
+
+			count := metrics.TypedSpec().Value.RegisteredMachinesCount
+			select {
+			case <-ctx.Done():
+				return nil
+			case countChan <- count:
+				// Successfully sent to the channel.
+			default:
+				log.Printf("Dropping count update: %d", count)
+			}
+		}
+	}
+}
+
+// debouncer accumulates count changes and updates Stripe after a debounce duration.
+func debouncer(ctx context.Context, countChan <-chan uint32, subscriptionItemID string) {
```

Comment:
I don't think I understand why `debouncer` is a separate goroutine, it hurts error reporting, but I think it can be perfectly merged into the main controller loop reducing complexity